docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,187 @@
+
+**In-tree plugin**: Plugins which are included in Kubernetes project/source tree.
+
+## API:
+The existing flex volume plugin driver API is in the alpha phase and are subject to change between Kubernetes versions. This section proposes a stable API for plugin driver writers to work with.
+
+### Proposal:
+Proposes fewer APIs with stable request & response API objects.
+
+###### Pros:
+* Has standard request & response API objects.
+* Supports versioning.
+* Uses gRPC.
+* Fewer calls.
+* Simple and clean interface.
+
+###### Cons:
+* New API objects.
+
+#### APIs
+
+```protobuf
+
+
+// Capabilities of the driver
+message DriverCapabilities {
+    enum ReadWriteMode {
+        READ_ONLY = 0;
+        READ_WRITE_ONCE = 1;
+        READ_WRITE_MANY = 2;
+    }
+	ReadWriteMode supported_read_write_mode = 1;
+    enum AttachmentPolicy {
+        NONE = 0;  // Driver does not support attach/detach.
+        LOCAL = 1; // Driver supports local volume attachments from kubelet.
+        REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+    }
+    AttachmentPolicy supported_attachment_policy = 2;
+    enum Capabilities {
+        DYNAMIC_PROVISIONING = 0;
+        CUSTOM_MOUNT = 1;
+        SELINUX = 2,
+        OWNERSHIP_MANAGEMENT = 3;
+        METRICS = 4;
+    }
+    repeated Capabilities supported_capabilities = 3;
+}
+
+message ProbeDriverRequest {}
+
+message ProbeDriverResponse {
+    DriverCapabilities capabilities = 1;
+    repeated string options = 2;
+}
+
+service FlexVolumeService {
+    rpc ProbeDriver(ProbeDriverRequest) returns (ProbeDriverResponse) {}
+    rpc Create(CreateRequest) returns (CreateResponse) {}
+    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+    rpc Attach(AttachRequest) returns (AttachResponse) {}
+    rpc Detach(DetachRequest) returns (DetachResponse) {}
+    rpc Mount(MountRequest) returns (MountResponse) {}
+    rpc Unmount(UnmountRequest) returns (UnmountResponse) {}
+    rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}
+}
+
+// Volume source.
+message Spec {
+    string name = 1; // Name of the volume.
+    string fstype = 2; // Requested file system type.
+    string read_write_mode = 3;
+    uint64 size = 4; // Size of the volume.
+    map<string, string> secrets = 5; // Secrets required for the driver to talk to its own controller.
+    map<string, string> options = 6; // Extra options passed to the driver.
+}
+
+message CreateRequest{
+    Spec spec = 1; // Volume specification.
+}
+
+message CreateResponse{
+    string id = 1; // Volume identifier, UUID/name
+}
+
+message DeleteRequest{
+    string name = 1; // Name of the volume.
+}
+
+message DeleteResponse{}
+
+message AttachRequest {
+    Spec spec = 1; // Volume spec.
+    string host = 2; // Name of the host to attach volume to.
+}
+
+message AttachResponse{
+    string device_path = 1; // Path to the device where the volume is attached.
```

Comment:
Please add a comment that `""` is allowed and should be interpreted as the driver does not know where the volume was attached at this point. It's up to the driver to find the device in `MountRequest`.